stl/inc/format

```diff
@@ -1187,6 +1187,24 @@ public:
         }
     }
 
+    _NODISCARD size_t _Calculate_reserve_size() const noexcept {
+        using _CharT        = typename _Context::char_type;
+        size_t _Result      = 0;
+        const auto _Visitor = [&_Result]<class _ArgTy>(const _ArgTy _Arg) noexcept {
+            if constexpr (is_same_v<_ArgTy, basic_string_view<_CharT>>) {
+                _Result += _Arg.size();
+            } else if constexpr (is_same_v<_ArgTy, const _CharT*>) {
+                _Result += char_traits<_CharT>::length(_Arg);

```

Comment:
Could this lead to determining the strlen multiple times? Do we cache it anywhere?

Reply:
I'm afraid it could since [_Write](https://github.com/microsoft/STL/blob/fc1b784ec863cfa91428384abcbc21bce27496ea/stl/inc/format#L1891-L1895) function is converting it to a `basic_string_view` later. Since caching it here would be problematic, we could convert all `const _CharT*` to `basic_string_view` in `make_(w)format_args`, although it would probably be non-conformant (but TBH the chance of anyone even noticing it would be extremely low, so I think we should do that (we can also file an LWG issue to allow us for it)).

Reply:
that conversion itself would run strlen, just pick a number

Reply:
> that conversion itself would run strlen

That's the point, we call `strlen` once in `make_(w)format_args`, we don't need to call it twice (once here, and once in `_Write`)

Reply:
The point of storing `const CharT*` as `const CharT*` instead of `basic_string_view` is that we can potentially avoid `strlen` entirely if the format spec specifies a maximum length.

Reply:
I might be missing something but it seems to me that `_Write` function for `const CharT*` is just calling `_Write` function for `basic_string_view`. https://github.com/microsoft/STL/blob/c6f02b6d59eb32baf17007798db4b23fc254699d/stl/inc/format#L1909-L1913

Reply:
Also is there a format spec for string that specifies maximum length?

Reply:
> I might be missing something but it seems to me that `_Write` function for `const CharT*` is just calling `_Write` function for `basic_string_view`.

We're not finished yet =)

> Also is there a format spec for string that specifies maximum length?

[[[format.string.std]/14](http://eel.is/c++draft/format.string.std#14):

The _nonnegative-integer_ in _precision_ is a decimal integer defining the precision or maximum field size. ... For string types, this field provides an upper bound for the estimated width of the prefix of the input string that is copied into the output.